### PR TITLE
Add conv sw test

### PR DESCRIPTION
**Pull request by xiaoling-yi**
In this PR, we add the hardware cfg and sw test for streamer-gemm dedicated for conv operation.
It instances more loops in the streamer.

### CODE REVIEW

**Review by rgantonio — APPROVED** [review]
Good job!!! ❤️ ❤️ ❤️ ❤️ ❤️

File: target/snitch_cluster/cfg/snax-streamer-gemm-conv.hjson [hunk 1]
```diff
@@ -0,0 +1,185 @@
+// Copyright 2023 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Cluster configuration for a simple testbench system.
```
Comment (JosseVanDelm) [inline target/snitch_cluster/cfg/snax-streamer-gemm-conv.hjson:5]: This is not a simple testbench system, is it?

File: target/snitch_cluster/cfg/snax-streamer-gemm-conv.hjson [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        num_int_outstanding_loads: 1,
+        num_int_outstanding_mem: 4,
+        num_fp_outstanding_loads: 4,
+        num_fp_outstanding_mem: 4,
+        num_sequencer_instructions: 16,
+        num_dtlb_entries: 1,
+        num_itlb_entries: 1,
+        // Enable division/square root unit
+        // Xdiv_sqrt: true,
+    },
```
Comment (JosseVanDelm) [inline target/snitch_cluster/cfg/snax-streamer-gemm-conv.hjson:128]: This template is not used

File: target/snitch_cluster/sw/snax/streamer-gemm-conv/src/snax-streamer-gemm-conv-lib.c [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    write_csr(997, tempLoop2);
+
+    // set subtraction a and b
+    write_csr(998, subtractions);
+}
+
+// Set CSR to start GEMM
+void set_conv_block_gemm_start() { write_csr(999, 1); }
+
+// Poll until Streamer and GEMM accelerator finish
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/snax/streamer-gemm-conv/src/snax-streamer-gemm-conv-lib.c:110]: Are these polling though?
```suggestion
// Stall until Streamer and GEMM accelerator finish
```

File: target/snitch_cluster/sw/apps/snax-streamer-gemm-conv/data/datagen.py [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+    # row major to column major for B
+    im2col_kernel = im2col_kernel.T
+    im2col_kernel = data_reshuffler_golden_model(
+        K, N, 8, 8, 8, 8 * 8 * K, 1, 8 * K, im2col_kernel.reshape(-1)
+    )
+    im2col_conv2d_res = block_gemm_golden_model(
+        M, K, N, 8, 8, 8, im2col_matrix, im2col_kernel, 0, 0
+    )
+    # im2col_conv2d_res = data_reshuffler_golden_model(N, M, 8, 8, 8, 8 * 8 * N,
+    # 1, 8 * N, im2col_conv2d_res, 1)
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-streamer-gemm-conv/data/datagen.py:411]: Why is this commented out?

File: target/snitch_cluster/sw/apps/snax-streamer-gemm-conv/data/datagen.py [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+
+    # Writing testing data and golden data into data.h
+    # implicit im2col matrix and kernel, store original input data and kernel
+    data_str += [format_vector_definition("int8_t", "A", input_padding.reshape(-1))]
+    data_str += [format_vector_definition("int8_t", "B", kernel.reshape(-1))]
+
+    # explicit im2col matrix and kernel, store the columned input data
+    # for comparing with the implicit im2col method
+    # data_str += [format_vector_definition("int8_t", "A", im2col_matrix)]
+    # data_str += [format_vector_definition("int8_t", "B", im2col_kernel)]
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-streamer-gemm-conv/data/datagen.py:426]: Why is this commented out?

File: target/snitch_cluster/sw/apps/snax-streamer-gemm-conv/data/datagen.py [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+    data_str += [
+        format_vector_definition("int32_t", "C_direct_conv2d", direct_conv2d_res)
+    ]
+
+    data_str = "\n\n".join(data_str)
+
+    return data_str
+
+
+def test():
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-streamer-gemm-conv/data/datagen.py:440]: What is this test for?

File: target/snitch_cluster/sw/apps/snax-streamer-gemm-conv/data/datagen.py [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+                    ] = data[
+                        # input address calculation with
+                        # strided layout mapping eqaution
+                        strides["M"] * M
+                        + strides["K"] * K
+                        + strides["m"] * m
+                        + strides["k"] * k
+                    ]
+
+    return result_array.ravel()
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-streamer-gemm-conv/data/datagen.py:194]: Is this not tested elsewhere, and can we thus not get that function from somewhere else?

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]
This looks like a very good PR! However, I really would need some extra help understanding what is going on and I think this PR requires a lot more documentation. I also don't have a lot of time to check it in this much detail, but I still think some important bits are missing that we would need.
Some things are general comments (can be solved in later PRs) other things are quite important to get right ASAP (this PR hopefully?)

* For a 1,152 PR the explanation is not that detailed. Can you please add a bit more context in the PR description? (This PR!)
* There is not a lot documentation going when it comes to data layout. Conv2D is very elaborate in its possiblities, NHWC, NCHW, NCHWc...
What did you pick here and why? (This PR, current assumption/initial thought is fine, please just document the currently assumed data layouts)
* Can you please do a really good effort in documenting what each of these parameters need to be set to. I'd rather have too much docs/comments here than too few.
There are many parameters to set here, and it takes just one to be misconfigured to lead to very long and annoying debugging in the end. (This PR please please please)
* Can you please document all the constraints this conv2d unit has? Does it support all kinds of strides, pads, or do left and right padding always need to be the same maybe? (Current assumptions and limitiations in this PR please!)
* While it is great that you are testing (Really ! Thanks!) You are testing this entire unit with just one set of parameters? How can we ever know it will not trigger some weird behaviour in some corner case? (Not this PR? Not sure?)
* What is an implicit im2col?
* General comment: I have the impression that a lot of this code is copy pasted and that this leads to a lot of code duplication. If we are copy pasting a lot that should ring some alarm bell, can we think of ways to add accelerators to this in less than 1152 lines?  (Not this PR?)

**Comment by JosseVanDelm** [comment]
Okay for me to put this PR through and delay the docs for later, but please do address some of the minor comments @xiaoling-yi

**Review by JosseVanDelm — APPROVED** [review]
Please still have a look at some of the suggestions! Thanks for this work @xiaoling-yi !

**Comment by xiaoling-yi** [comment]
Hi @JosseVanDelm thanks for your comments! I address some! The repeated golden model codes will be put in a python util in later PR when I have more time! I will add the doc later after the tight tape out ddl..
Let's leave the comments for future reference!

**Pull request merged**